osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerReadyButton.cs

```diff
@@ -55,16 +59,150 @@ private void load(GameHost host, AudioManager audio)
                 RulesetID = Beatmap.Value.BeatmapInfo.Ruleset.OnlineID
             };
 
-            if (button != null)
-                Remove(button);
+            Child = new PopoverContainer
+            {
+                RelativeSizeAxes = Axes.Both,
+                Child = button = new MultiplayerReadyButton
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Size = new Vector2(200, 50),
+                }
+            };
+        });
 
-            Add(button = new MultiplayerReadyButton
+        [Test]
+        public void TestStartWithCountdown()
+        {
+            ClickButtonWhenEnabled<MultiplayerReadyButton.ReadyButton>();
+            AddUntilStep("countdown button shown", () => this.ChildrenOfType<MultiplayerReadyButton.CountdownButton>().SingleOrDefault()?.IsPresent == true);
+            ClickButtonWhenEnabled<MultiplayerReadyButton.CountdownButton>();
+            AddStep("click the first countdown button", () =>
             {
-                Anchor = Anchor.Centre,
-                Origin = Anchor.Centre,
-                Size = new Vector2(200, 50),
+                var popoverButton = this.ChildrenOfType<MultiplayerReadyButton.CountdownButton.PopoverButton>().First();
+                InputManager.MoveMouseTo(popoverButton);
+                InputManager.Click(MouseButton.Left);
             });
-        });
+
+            AddAssert("countdown button not visible", () => !this.ChildrenOfType<MultiplayerReadyButton.CountdownButton>().Single().IsPresent);
+            AddStep("finish countdown", () => MultiplayerClient.FinishCountDown());
```

Comment:
This is fine for testing the button itself, but there doesn't seem to be any tests covering the actual countdown flow (ie. setting the countdown for 1 second and then waiting 1 second until the game starts, something similar to `TestGameplayFlow`).

Reply:
That wasn't the intention to begin with, because that would be testing test behaviour - `TestMultiplayerClient` is implemented mostly to emulate the server, but it isn't exact, as noted by the distinct lack of sanity checks it does which the server tests for instead.  
That's to say, if this were to be tested, it would be done in osu-server-spectator, not here.

The point of this test is to make sure that the ready button sends the countdown start and stop requests.

Reply:
osu-server-spectator test added here: https://github.com/ppy/osu-server-spectator/pull/109/commits/a148f666ceb9803f57748f8fe132ead0833bf2cd